[OSDOCS-5240] Installer support to automatically create the MachineSets when installing in existing VPC on AWS w/Local Zones (#57427)
openshift/openshift-docs: modules/machines-edge-machine-pool.adoc

> 
> :_content-type: CONCEPT

Author: ```suggestion
// Module included in the following assemblies
// * installing/installing_aws/installing-aws-localzone.adoc

:_content-type: CONCEPT
```

[id="machines-edge-machine-pool_{context}"]
= The edge compute pool for AWS Local Zones

{product-title} 4.12 introduced a new compute pool, _edge_, that is designed for use in remote zones. The edge compute pool configuration is common between AWS Local Zone locations. However, due to the type and size limitations of resources like EC2 and EBS on Local Zone resources, the default instance type that is created can vary from the traditional worker pool.

> The default Elastic Block Store (EBS) for Local Zone locations is `gp2`, which differs from the regular worker pool. The instance type used for each Local Zone on edge compute pool also may differ from worker pools, depending on the instance offerings on the zone.

Author: ```suggestion
The default Elastic Block Store (EBS) for Local Zone locations is `gp2`, which differs from the regular worker pool. The instance type used for each Local Zone on edge compute pool also might differ from worker pools, depending on the instance offerings on the zone.
```

> The default elastic block store (EBS) for Local Zone locations is `gp2`, which differs from the default worker pool. The preferred list of instance types for the edge compute pool follows the same order of worker pools. Depending on the availability of the location, the system chooses one of those instances. The preferred list of instance types for the edge compute pool includes:
> 
> * `m6i.xlarge`
> * `m5.xlarge`
> * `c5d.2xlarge`

Reviewer: I think we can stop on `Depending on the availability of the location`.

This list, order, or logic can be changed over time on the installer, let's avoid maintenance on the docs. 

```suggestion
The default elastic block store (EBS) for Local Zone locations is `gp2`, which differs from the default worker pool. The preferred list of instance types for the edge compute pool follows the same order of worker pools, depending on the availability of the location.
```

cc @yunjiang29 (wdyt?)

Reviewer 2: @mtulio What is the difference between this section and https://github.com/openshift/openshift-docs/pull/57057, I think maintaining one list is enough.

* `machine.openshift.io/zone-type=local-zone`

> * `machine.openshift.io/zone-group=<$ZONE_GROUP_NAME>`

Reviewer 2: Remove `<>`?
```suggestion
* `machine.openshift.io/zone-group=$ZONE_GROUP_NAME`
```

> 

Author: ```suggestion
```

* `node-role.kubernetes.io/edge=''`
* `zone_type=local-zone`

> * `zone_group=<LOCAL_ZONE_GROUP>`

Reviewer: Those labels should be updated to:

- `machine.openshift.io/zone-group=<$ZONE_GROUP_NAME>`
- `machine.openshift.io/zone-type=local-zone`

Reviewer: Could you please make sure those labels are set correctly on the entire Local Zone book? It's defined on the Enhancement Proposal and installer implementation for phase-1, but it seems to passed from the phase-0 docs review.

Author: Beyond what this PR touches, I'm not seeing these labels anywhere.

> By default, the system creates the edge compute pool only if users add AWS Local Zone subnet IDs to the list `platform.aws.subnets`.

Reviewer: This will be changed on 4.14 and we will support to create subnets when the zone will be added on the edge pool.

I would suggest reviewing that phrase to something like: "By default, the installer creates the edge compute pool manifests when installing in existing VPC, only when the subnet ID is added to the list `platform.aws.subnets`.

- Phase-1 (this, 4.13): create edge manifests when installing in existing VPC
- Phase-2 (plan, 4.14): create the zone name for Local Zones is specified in the edge compute pool

```suggestion
By default, the installer creates the edge compute pool manifests only if users add AWS Local Zone subnet IDs to the list `platform.aws.subnets` on install-config.yaml.
```

Reviewer: > the installer creates the edge compute pool manifests 

installer will create **the manifests**, not the pool itself. I think stating this could provide more understanding what the installer is doing. wdyt?

Author: Fair. 

```suggestion
By default, the system creates the edge compute pool manifest only if users add AWS Local Zone subnet IDs to the list `platform.aws.subnets`.
```

> The edge compute pool's machine sets have a `NoSchedule taint` by default to prevent regular workloads from being spread out on those machines. Users can only run user workloads if the tolerations are defined on the pod spec. This practice is demonstrated in the following examples:

Reviewer: The examples below do not match this statement.

The examples are from `install-config.yaml`.

I think you can remove the last part of this phrase and create a new paragraph summarizing/introducing the examples below. wdyt?

Author: Works for me. Paragraph aside, do those examples all seem valuable to you? @mtulio

Reviewer: if you would like to provide a full example, it's also fine for me. :)

[source,yaml]
----
apiVersion: v1
baseDomain: devcluster.openshift.com
metadata:
  name: ipi-localzone
platform:
  aws:
    region: us-west-2
    subnets:

>       - subnetID-private1
>       - subnetID-private2
>       - subnetID-private3
>       - subnetID-public1
>       - subnetID-public2
>       - subnetID-public3
>       - subnetID-localzone-public1

Author: ```suggestion
   - publicSubnetId-1
    - publicSubnetId-2
    - publicSubnetId-3
    - privateSubnetId-1
    - privateSubnetId-2
    - privateSubnetId-3
      - subnetID-localzone-public1
```

@yunjiang29 ?

Author: ```suggestion
      - publicSubnetId-1
      - publicSubnetId-2
      - publicSubnetId-3
      - privateSubnetId-1
      - privateSubnetId-2
      - privateSubnetId-3
      - publicSubnetId-LocalZone-1
```

    - subnet-093f00deb44ce81f4

>     - subnet-0f85ae65796e8d107

Reviewer: Considering this example is related to the CloudFormation exemplified above which creates subnets (public and private) only in 3 AZs: do you think didactic we could replace this with something like:
```
    - subnetID-private1
    - subnetID-private2
    - subnetID-private3
    - subnetID-public1
    - subnetID-public2
    - subnetID-public3
    - subnetID-localzone-public1
```

----

.Configuration that uses an edge pool with a custom instance type
[source,yaml]
----
apiVersion: v1
baseDomain: devcluster.openshift.com
metadata:
  name: ipi-localzone
compute:
- name: edge
  platform:
    aws:
      type: m5.4xlarge
platform:
  aws:
    region: us-west-2
    subnets:

>     - subnet-0fc845d8e30fdb431
>     - subnet-0a2675b7cbac2e537
>     - subnet-01c0ac400e1920b47
>     - subnet-0fee60966b7a93da6
>     - subnet-002b48c0a91c8c641
>     - subnet-093f00deb44ce81f4
>     - subnet-0f85ae65796e8d107

Author: ```suggestion
      - publicSubnetId-1
      - publicSubnetId-2
      - publicSubnetId-3
      - privateSubnetId-1
      - privateSubnetId-2
      - privateSubnetId-3
      - subnetID-localzone-public1
```

@yunjiang29 ?

Reviewer: @maxwelldb ACK for @yunjiang29's suggestion, fixing the Local Zone subnet to:
- `- publicSubnetId-LocalZone-1`

Author: ```suggestion
      - publicSubnetId-1
      - publicSubnetId-2
      - publicSubnetId-3
      - privateSubnetId-1
      - privateSubnetId-2
      - privateSubnetId-3
      - publicSubnetId-LocalZone-1
```
@mtulio Like so?

    - subnet-093f00deb44ce81f4
    - subnet-0f85ae65796e8d107
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
----

> .Example edge pool with a custom EBS type

Reviewer: ```suggestion
.Example edge pool with a custom EBS type `gp3`
```

> Instance types differ between locations. Check the AWS documentation to verify availability in the Local Zone in which the cluster will run.

Author: ```suggestion
Instance types differ between locations. To verify availability in the Local Zone in which the cluster will run, see the AWS documentation.
```

[source,yaml]
----
apiVersion: v1
baseDomain: devcluster.openshift.com
metadata:
  name: ipi-localzone
compute:
- name: edge
  platform:
    aws:
      rootVolume:

>         type: gp3

Reviewer 2: @mtulio @maxwelldb change to `gp2`? as `gp3` is not supported

Author: ```suggestion
        type: gp2
```
@mtulio Any objections?

Reviewer: @yunjiang29 `gp3`, there are zones that support `gp3`, two specifically: `us-west-2-lax-1{a,b}. It can change over the time, the goal here is to provide an example how to replace the default, `gp2`.

Do you think it may be confusing? I am thinking to aggregate all those examples to keep it simple and standardized with our docs. WDYT?

Author: @yunjiang29 @mtulio Any decision on this?

Reviewer: Keep `gp3`, it's intentional to keep it here exemplifying how to change the default (`gp2`).
`gp3` is supported in some Locations: https://aws.amazon.com/about-aws/global-infrastructure/localzones/features/

Reviewer 2: > Is it confusing? I am thinking of aggregating all those examples to keep it simple and standardized with our docs. WDYT?

> Keep gp3

@mtulio @maxwelldb  it looks good to me
